src/dune_rules/pkg_rules.ml

```diff
@@ -1726,26 +1934,38 @@ let setup_package_rules context ~dir ~pkg_name : Gen_rules.result Memo.t =
 ;;
 
 let setup_rules ~components ~dir ctx =
+  let open Rule_root in
   match components with
+  | [ ".ocamlformat"; pkg_name ] ->
+    setup_package_rules ctx ~component:(Dev_tool Ocamlformat) ~dir ~pkg_name
+  | [ ".ocamlformat" ] ->
```

Comment:
Why do you need to introduce a special directory just for ocamlformat?

Reply:
To avoid conflict with `dune.lock` packages. Like the `ocamlformat` could have the same dependency with different version, so it is built separately.

Reply:
If you indeed have two versions of ocamlformat, which one are you going to let the user see then?

Reply:
The priority is the one  described `dune-project`.